# LAB ASSIGNMENT 2 - Sentiment Analysis of Amazon Review

Group Members:-

Muhammad Fahmi Bin Misri (SW01081019)
Nik Muhammad Nafis Bin Nik Azlan (SW01081028)

In [1]:
import pandas as pd
import re
import string 
import nltk
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer

In [2]:
df = pd.read_csv("Reviews.csv")
df.head()

Id   ProductId          UserId                      ProfileName  \
0   1  B001E4KFG0  A3SGXH7AUHU8GW                       delmartian   
1   2  B00813GRG4  A1D87F6ZCVE5NK                           dll pa   
2   3  B000LQOCH0   ABXLMWJIXXAIN  Natalia Corres "Natalia Corres"   
3   4  B000UA0QIQ  A395BORC6FGVXV                             Karl   
4   5  B006K2ZZ7K  A1UQRSCLF8GW1T    Michael D. Bigham "M. Wassir"   

   HelpfulnessNumerator  HelpfulnessDenominator  Score        Time  \
0                     1                       1      5  1303862400   
1                     0                       0      1  1346976000   
2                     1                       1      4  1219017600   
3                     3                       3      2  1307923200   
4                     0                       0      5  1350777600   

                 Summary                                               Text  
0  Good Quality Dog Food  I have bought several of the Vitality canned d...  
1      Not as Advertised  Product arrived labeled as Jumbo Salted Peanut...  
2  "Delight" says it all  This is a confection that has been around a fe...  
3         Cough Medicine  If you are looking for the secret ingredient i...  
4            Great taffy  Great taffy at a great price.  There was a wid...

In [3]:
df.shape

(568454, 10)

In [4]:
# read data - first 1000 rows
df = pd.read_csv("Reviews.csv", nrows=1000)

In [5]:
df.head(10)

Id   ProductId          UserId                      ProfileName  \
0   1  B001E4KFG0  A3SGXH7AUHU8GW                       delmartian   
1   2  B00813GRG4  A1D87F6ZCVE5NK                           dll pa   
2   3  B000LQOCH0   ABXLMWJIXXAIN  Natalia Corres "Natalia Corres"   
3   4  B000UA0QIQ  A395BORC6FGVXV                             Karl   
4   5  B006K2ZZ7K  A1UQRSCLF8GW1T    Michael D. Bigham "M. Wassir"   
5   6  B006K2ZZ7K   ADT0SRK1MGOEU                   Twoapennything   
6   7  B006K2ZZ7K  A1SP2KVKFXXRU1                David C. Sullivan   
7   8  B006K2ZZ7K  A3JRGQVEQN31IQ               Pamela G. Williams   
8   9  B000E7L2R4  A1MZYO9TZK0BBI                         R. James   
9  10  B00171APVA  A21BT40VZCCYT4                    Carol A. Reed   

   HelpfulnessNumerator  HelpfulnessDenominator  Score        Time  \
0                     1                       1      5  1303862400   
1                     0                       0      1  1346976000   
2                     1                       1      4  1219017600   
3                     3                       3      2  1307923200   
4                     0                       0      5  1350777600   
5                     0                       0      4  1342051200   
6                     0                       0      5  1340150400   
7                     0                       0      5  1336003200   
8                     1                       1      5  1322006400   
9                     0                       0      5  1351209600   

                                         Summary  \
0                          Good Quality Dog Food   
1                              Not as Advertised   
2                          "Delight" says it all   
3                                 Cough Medicine   
4                                    Great taffy   
5                                     Nice Taffy   
6  Great!  Just as good as the expensive brands!   
7                         Wonderful, tasty taffy   
8                                     Yay Barley   
9                               Healthy Dog Food   

                                                Text  
0  I have bought several of the Vitality canned d...  
1  Product arrived labeled as Jumbo Salted Peanut...  
2  This is a confection that has been around a fe...  
3  If you are looking for the secret ingredient i...  
4  Great taffy at a great price.  There was a wid...  
5  I got a wild hair for taffy and ordered this f...  
6  This saltwater taffy had great flavors and was...  
7  This taffy is so good.  It is very soft and ch...  
8  Right now I'm mostly just sprouting this so my...  
9  This is a very healthy dog food. Good for thei...

In [6]:
df.shape

(1000, 10)

In [7]:
# select Text column
df_t = df['Text']
df_t.head()

0    I have bought several of the Vitality canned d...
1    Product arrived labeled as Jumbo Salted Peanut...
2    This is a confection that has been around a fe...
3    If you are looking for the secret ingredient i...
4    Great taffy at a great price.  There was a wid...
Name: Text, dtype: object

In [8]:
# check no. of duplicates in Text column
df_t_duplicates = df_t.duplicated()
print(df_t_duplicates.sum())

3


In [9]:
# remove duplicates in Text column
df_rdup = df_t.drop_duplicates(keep='first')
df_rdup.shape

(997,)

Removing HTML tags and unwanted characters

In [10]:
def get_cleaned_textdata(sentence):
    modified_sentence = re.sub(r'<.*?>',' ', sentence)
    modified_sentence = ''.join([i if i not in string.punctuation else ' ' for i in modified_sentence])
    modified_sentence = re.sub(r'\d+', ' ', modified_sentence)
    modified_sentence = re.sub(r'\s+', ' ', modified_sentence)
    modified_sentence = modified_sentence.lower()
    return modified_sentence

In [11]:
df_rdupclean = df_rdup.apply(get_cleaned_textdata)

In [12]:
# check any row
print(df_rdupclean[169])

the item over all was fine but the banana heads are not just like banana runts they are alot smaller also when i received the item the packaging was not great the banana heads had come open during shipping and were all over the packing envelope 


In [13]:
print(df_rdupclean)

0      i have bought several of the vitality canned d...
1      product arrived labeled as jumbo salted peanut...
2      this is a confection that has been around a fe...
3      if you are looking for the secret ingredient i...
4      great taffy at a great price there was a wide ...
                             ...                        
995    black market hot sauce is wonderful my husband...
996    man what can i say this salsa is the bomb i ha...
997    this sauce is so good with just about anything...
998    not hot at all like the other low star reviewe...
999    i have to admit i was a sucker for the large q...
Name: Text, Length: 997, dtype: object


In [14]:
# Select relevant columns for sentiment analysis
df = df[['Score', 'Text']]

Tokenizing Text

In [15]:
nltk.download('punkt')

# Tokenize the text into individual words
df['Tokens'] = df['Text'].apply(lambda x: nltk.word_tokenize(x))

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\Fahmi\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!


In [16]:
# Download the WordNet lemmatizer
nltk.download('wordnet')

# Initialize the WordNet lemmatizer
lemmatizer = WordNetLemmatizer()

# Lemmatize the tokens
df['Tokens'] = df['Tokens'].apply(lambda x: [lemmatizer.lemmatize(word) for word in x])

[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\Fahmi\AppData\Roaming\nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


In [17]:
# Join the tokens back into sentences
df['Preprocessed_Text'] = df['Tokens'].apply(lambda x: ' '.join(x))

# Save the preprocessed data to a new CSV file
df.to_csv('preprocessed_amazon_reviews.csv', index=False)

# Preview the preprocessed data
df.head()

Score                                               Text  \
0      5  I have bought several of the Vitality canned d...   
1      1  Product arrived labeled as Jumbo Salted Peanut...   
2      4  This is a confection that has been around a fe...   
3      2  If you are looking for the secret ingredient i...   
4      5  Great taffy at a great price.  There was a wid...   

                                              Tokens  \
0  [I, have, bought, several, of, the, Vitality, ...   
1  [Product, arrived, labeled, a, Jumbo, Salted, ...   
2  [This, is, a, confection, that, ha, been, arou...   
3  [If, you, are, looking, for, the, secret, ingr...   
4  [Great, taffy, at, a, great, price, ., There, ...   

                                   Preprocessed_Text  
0  I have bought several of the Vitality canned d...  
1  Product arrived labeled a Jumbo Salted Peanuts...  
2  This is a confection that ha been around a few...  
3  If you are looking for the secret ingredient i...  
4  Great taffy at a great price . There wa a wide...

# FEATURE EXTRACTION (BAG OF WORDS AND TF-IDF)

In [18]:
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer

Bag Of Words

In [19]:
# Initialize the CountVectorizer
vectorizer = CountVectorizer()

# Fit and transform the preprocessed text data
bow_features = vectorizer.fit_transform(df['Preprocessed_Text'])

# Get the vocabulary (unique words)
vocabulary = vectorizer.get_feature_names_out()

# Print the shape of the BoW features and the vocabulary size
print("BoW feature shape:", bow_features.shape)
print("Vocabulary size:", len(vocabulary))

BoW feature shape: (1000, 5549)
Vocabulary size: 5549


In [20]:
from collections import Counter

# Given list of sentences
sentences = [
    "product arrived labeled as jumbo salted peanuts the peanuts were actually small sized unsalted not sure if this was an error or if the vendor intended to represent the product as jumbo"
    "if you are looking for the secret ingredient in robitussin i believe i have found it i got this in addition to the root beer extract i ordered which was good and made some cherry soda the flavor is very medicinal",
    "great taffy at a great price there was a wide assortment of yummy taffy delivery was very quick if your a taffy lover this is a deal",
    "this taffy is so good it is very soft and chewy the flavors are amazing i would definitely recommend you buying it very satisfying"
]

# Step 1: List all unique words
all_words = ' '.join(sentences).split()
unique_words = list(set(all_words))

# Step 2: Create word frequency table
word_freq = Counter(all_words)
sorted_word_freq = dict(sorted(word_freq.items(), key=lambda item: item[1], reverse=True))

# Step 3: Create document vectors
document_vectors = []
for sentence in sentences:
    sentence_words = sentence.split()
    document_vector = [1 if word in sentence_words else 0 for word in sorted_word_freq.keys()]
    document_vectors.append(document_vector)

# Print the results
print("Step 1: List of Unique Words:", unique_words)
print("\nStep 2: Word Frequency Table:")
print("Word\tFrequency")
for word, freq in sorted_word_freq.items():
    print(f"{word}\t{freq}")

print("\nStep 3: Document Vectors:")
for sentence, vector in zip(sentences, document_vectors):
    print(f"Sentence: {sentence}\nVector: {vector}\n")

Step 1: List of Unique Words: ['a', 'this', 'the', 'were', 'intended', 'product', 'robitussin', 'quick', 'are', 'so', 'unsalted', 'soda', 'assortment', 'amazing', 'you', 'salted', 'ingredient', 'deal', 'as', 'flavors', 'if', 'medicinal', 'flavor', 'actually', 'sure', 'error', 'good', 'which', 'in', 'believe', 'is', 'there', 'satisfying', 'addition', 'vendor', 'great', 'i', 'wide', 'found', 'made', 'for', 'extract', 'represent', 'labeled', 'chewy', 'have', 'secret', 'looking', 'or', 'peanuts', 'it', 'sized', 'lover', 'delivery', 'jumboif', 'price', 'your', 'arrived', 'very', 'small', 'to', 'was', 'of', 'root', 'and', 'jumbo', 'recommend', 'cherry', 'not', 'yummy', 'at', 'would', 'definitely', 'buying', 'taffy', 'some', 'soft', 'an', 'got', 'ordered', 'beer']

Step 2: Word Frequency Table:
Word	Frequency
the	7
i	5
this	4
was	4
is	4
very	4
taffy	4
a	4
if	3
it	3
product	2
as	2
peanuts	2
to	2
you	2
are	2
in	2
good	2
and	2
great	2
arrived	1
labeled	1
jumbo	1
salted	1
were	1
actually	1
small	

Term Frequency-Inverse Document Frequency (TF-IDF)

In [21]:
# Initialize the TfidfVectorizer
tfidf_vectorizer = TfidfVectorizer()

# Fit and transform the preprocessed text data
tfidf_features = tfidf_vectorizer.fit_transform(df['Preprocessed_Text'])

# Get the vocabulary (unique words)
tfidf_vocabulary = tfidf_vectorizer.get_feature_names_out()

# Print the shape of the TF-IDF features and the vocabulary size
print("TF-IDF feature shape:", tfidf_features.shape)
print("Vocabulary size:", len(tfidf_vocabulary))

TF-IDF feature shape: (1000, 5549)
Vocabulary size: 5549


In [22]:
import math
from collections import Counter

def compute_tf(document):
    word_count = Counter(document)
    tf = {word: count/len(document) for word, count in word_count.items()}
    return tf

def compute_idf(documents):
    N = len(documents)
    idf = {}
    all_words = set(word for doc in documents for word in doc)
    for word in all_words:
        count = sum(1 for doc in documents if word in doc)
        idf[word] = math.log(N/count)
    return idf

def compute_tfidf(document, idf):
    tfidf = {}
    tf = compute_tf(document)
    for word, tf_value in tf.items():
        tfidf[word] = tf_value * idf[word]
    return tfidf

# New data
data = [
    "great taffy at a great price there was a wide assortment of yummy taffy delivery was very quick if your a taffy lover this is a deal",
    "this taffy is so good it is very soft and chewy the flavors are amazing i would definitely recommend you buying it very satisfying"
]

# Split data into tokens
documents = [doc.split() for doc in data]

# Compute TF for each document
tf_data = [compute_tf(doc) for doc in documents]

# Create DataFrame for TF
tf_df = pd.DataFrame(tf_data).fillna(0)
print("TF Scores:")
print(tf_df)

# Compute IDF
idf = compute_idf(documents)
idf_df = pd.DataFrame([idf]).fillna(0)
print("\nIDF Scores:")
print(idf_df)

# Compute TF-IDF for each document
tfidf_data = [compute_tfidf(doc, idf) for doc in documents]

# Create DataFrame for TF-IDF
tfidf_df = pd.DataFrame(tfidf_data).fillna(0)
print("\nTF-IDF Scores:")
print(tfidf_df)

TF Scores:
      great     taffy        at         a     price     there       was  \
0  0.074074  0.111111  0.037037  0.148148  0.037037  0.037037  0.074074   
1  0.000000  0.041667  0.000000  0.000000  0.000000  0.000000  0.000000   

       wide  assortment        of  ...   flavors       are   amazing  \
0  0.037037    0.037037  0.037037  ...  0.000000  0.000000  0.000000   
1  0.000000    0.000000  0.000000  ...  0.041667  0.041667  0.041667   

          i     would  definitely  recommend       you    buying  satisfying  
0  0.000000  0.000000    0.000000   0.000000  0.000000  0.000000    0.000000  
1  0.041667  0.041667    0.041667   0.041667  0.041667  0.041667    0.041667  

[2 rows x 37 columns]

IDF Scores:
       wide     price         a      your  this   flavors       the  very  \
0  0.693147  0.693147  0.693147  0.693147   0.0  0.693147  0.693147   0.0   

         if       was  ...   is     there      soft  satisfying      deal  \
0  0.693147  0.693147  ...  0.0  0.693147

# Model Selection (LEXICON APPROACH)

In [23]:
from textblob import TextBlob
from vaderSentiment.vaderSentiment import SentimentIntensityAnalyzer
from tabulate import tabulate
from nltk.sentiment import SentimentIntensityAnalyzer
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.svm import LinearSVC
from sklearn.metrics import accuracy_score, classification_report

1. Lexicon-based approach using NLTK's VADER (Valence Aware Dictionary and sEntiment Reasoner):

In [24]:
# Download the VADER lexicon
nltk.download('vader_lexicon')

# Assign sentiment labels based on the 'Score' column
def assign_sentiment(score):
    if score >= 4:
        return 'Positive'
    elif score <= 2:
        return 'Negative'
    else:
        return 'Neutral'

df['Sentiment'] = df['Score'].apply(assign_sentiment)

# Initialize the sentiment analyzer
sid = SentimentIntensityAnalyzer()

# Calculate sentiment scores for each review
df['Lexicon_Sentiment'] = df['Preprocessed_Text'].apply(lambda x: sid.polarity_scores(x)['compound'])

# Map sentiment scores to labels
df['Lexicon_Sentiment_Label'] = df['Lexicon_Sentiment'].apply(lambda x: 'Positive' if x > 0 else ('Negative' if x < 0 else 'Neutral'))

# Evaluate the lexicon-based approach
lexicon_accuracy = accuracy_score(df['Sentiment'], df['Lexicon_Sentiment_Label'])
print("Lexicon-based Approach Accuracy:", lexicon_accuracy)

[nltk_data] Downloading package vader_lexicon to
[nltk_data]     C:\Users\Fahmi\AppData\Roaming\nltk_data...
[nltk_data]   Package vader_lexicon is already up-to-date!


Lexicon-based Approach Accuracy: 0.804


In [25]:
data = [
    ("good flavor these came securely packed they were fresh and delicious i love these twizzlers", 'positive'),
    ("this is the same stuff you can buy at the big box stores there is nothing healthy about it it is just carbs and sugars save your money and get something that at least has some taste", 'negative'),
    ("the flavors are good however i do not see any differce between this and oaker oats brand they are both mushy", 'neutral'),
    ("the strawberry twizzlers are my guilty pleasure yummy six pounds will be around for a while with my son and i", 'positive'),
    ("this oatmeal is not good its mushy soft i don t like it quaker oats is the way to go", 'negative'),
    ("halloween is over but i sent a bag to my daughters class for her share the chocolate was fresh and enjoyed by many", 'neutral'),
    ("my daughter loves twizzlers and this shipment of six pounds really hit the spot it s exactly what you would expect six packages of strawberry twizzlers", 'positive'),
    ("arrived in days and were so stale i could not eat any of the bags", 'negative'),
    ("this bag of candy online is pretty expensive it should be cheaper in order to compete with grocery stores other than that its a good combination of my favorite candy", 'neutral'),
    ("i love eating them and they are good for watching tv and looking at movies it is not too sweet i like to transfer them to a zip lock baggie so they stay fresh so i can take my time eating them", 'positive'),
    ("no tea flavor at all just whole brunch of artifial flavors it is not returnable i wasted bucks", 'negative'),
    ("this stuff really works for preventing cramping during the middle to latter stages of your rides pop into each water bottle and you re set flavor is fine and goes down easy", 'neutral'),
    ("i fed this to my golden retriever and he hated it he wouldn t eat it and when he did it gave him terrible diarrhea we will not be buying this again it s also super expensive", 'negative'),
    ("i am very satisfied with my twizzler purchase i shared these with others and we have all enjoyed them i will definitely be ordering more", 'positive'),
    ("it is okay i would not go out of my way to buy it again", 'neutral'),
    ("candy was delivered very fast and was purchased at a reasonable price i was home bound and unable to get to a store so this was perfect for me", 'positive'),
    ("not what i was expecting in terms of the company s reputation for excellent home delivery products", 'negative'),
    ("this is the same food we get at pet store but it s delivered to my door and for the same price or slightly less ", 'neutral'),
    ("i can remember buying this candy as a kid and the quality hasn t dropped in all these years still a superb product you won t be disappointed with", 'positive'),
    ("i was disappointed in the flavor and texture of this mix i usually like most of the low carb things i have tried but was diappointed in this specific one ", 'negative'),
    ("i ve been eating ramen noodles since i was a little kid and i ve never found a better flavor than hot spicy chicken it isn t hot at all to a chilihead like me but it sure is good", 'neutral'),
    ("this was sooooo deliscious but too bad i ate em too fast and gained pds my fault", 'positive'),
    ("besides being smaller than runts they look the same and have the same consistency unfortunately they taste nothing like banana runts nor do they even taste good yucky stuff trying to return with vendor", 'negative'),
    ("the item over all was fine but the banana heads are not just like banana runts they are alot smaller also when i received the item the packaging was not great the banana heads had come open during shipping and were all over the packing envelope", 'neutral'),
]


In [26]:
table_data = [["Text", "Actual Label", "TextBlob Polarity", "TextBlob Sentiment", "VADER Compound", "VADER Sentiment"]]

In [27]:
for text, actual_label in data:
    # TextBlob
    blob = TextBlob(text)
    tb_polarity = blob.sentiment.polarity
    
    # Determine label based on polarity score from TextBlob
    if tb_polarity > 0:
        tb_label = 'positive'
    elif tb_polarity < 0:
        tb_label = 'negative'
    else:
        tb_label = 'neutral'

    # VADER
    analyzer = SentimentIntensityAnalyzer()
    vs = analyzer.polarity_scores(text)
    vader_compound = vs['compound']

    # Determine label based on compound score from VADER
    if vader_compound > 0.05:
        vader_label = 'positive'
    elif vader_compound < -0.05:
        vader_label = 'negative'
    else:
        vader_label = 'neutral'

    table_data.append([text, actual_label, tb_polarity, tb_label, vader_compound, vader_label])


In [28]:
print(tabulate(table_data, headers="firstrow", tablefmt="plain"))

Text                                                                                                                                                                                                                                                  Actual Label      TextBlob Polarity  TextBlob Sentiment      VADER Compound  VADER Sentiment
good flavor these came securely packed they were fresh and delicious i love these twizzlers                                                                                                                                                           positive                 0.58        positive                        0.9382  positive
this is the same stuff you can buy at the big box stores there is nothing healthy about it it is just carbs and sugars save your money and get something that at least has some taste                                                                 negative                 0.05        positive                        0.2363  positi

In [29]:
from textblob import TextBlob
from vaderSentiment.vaderSentiment import SentimentIntensityAnalyzer
from sklearn.metrics import classification_report
from tabulate import tabulate

data = [
    ("good flavor these came securely packed they were fresh and delicious i love these twizzlers", 'positive'),
    ("this is the same stuff you can buy at the big box stores there is nothing healthy about it it is just carbs and sugars save your money and get something that at least has some taste", 'negative'),
    ("the flavors are good however i do not see any differce between this and oaker oats brand they are both mushy", 'neutral'),
    ("the strawberry twizzlers are my guilty pleasure yummy six pounds will be around for a while with my son and i", 'positive'),
    ("this oatmeal is not good its mushy soft i don t like it quaker oats is the way to go", 'negative'),
    ("halloween is over but i sent a bag to my daughters class for her share the chocolate was fresh and enjoyed by many", 'neutral'),
    ("my daughter loves twizzlers and this shipment of six pounds really hit the spot it s exactly what you would expect six packages of strawberry twizzlers", 'positive'),
    ("arrived in days and were so stale i could not eat any of the bags", 'negative'),
    ("this bag of candy online is pretty expensive it should be cheaper in order to compete with grocery stores other than that its a good combination of my favorite candy", 'neutral'),
    ("i love eating them and they are good for watching tv and looking at movies it is not too sweet i like to transfer them to a zip lock baggie so they stay fresh so i can take my time eating them", 'positive'),
    ("no tea flavor at all just whole brunch of artifial flavors it is not returnable i wasted bucks", 'negative'),
    ("this stuff really works for preventing cramping during the middle to latter stages of your rides pop into each water bottle and you re set flavor is fine and goes down easy", 'neutral'),
    ("i fed this to my golden retriever and he hated it he wouldn t eat it and when he did it gave him terrible diarrhea we will not be buying this again it s also super expensive", 'negative'),
    ("i am very satisfied with my twizzler purchase i shared these with others and we have all enjoyed them i will definitely be ordering more", 'positive'),
    ("it is okay i would not go out of my way to buy it again", 'neutral'),
    ("candy was delivered very fast and was purchased at a reasonable price i was home bound and unable to get to a store so this was perfect for me", 'positive'),
    ("not what i was expecting in terms of the company s reputation for excellent home delivery products", 'negative'),
    ("this is the same food we get at pet store but it s delivered to my door and for the same price or slightly less ", 'neutral'),
    ("i can remember buying this candy as a kid and the quality hasn t dropped in all these years still a superb product you won t be disappointed with", 'positive'),
    ("i was disappointed in the flavor and texture of this mix i usually like most of the low carb things i have tried but was diappointed in this specific one ", 'negative'),
    ("i ve been eating ramen noodles since i was a little kid and i ve never found a better flavor than hot spicy chicken it isn t hot at all to a chilihead like me but it sure is good", 'neutral'),
    ("this was sooooo deliscious but too bad i ate em too fast and gained pds my fault", 'positive'),
    ("besides being smaller than runts they look the same and have the same consistency unfortunately they taste nothing like banana runts nor do they even taste good yucky stuff trying to return with vendor", 'negative'),
    ("the item over all was fine but the banana heads are not just like banana runts they are alot smaller also when i received the item the packaging was not great the banana heads had come open during shipping and were all over the packing envelope", 'neutral'),
]


# Initialize an empty list to store the data in tabular format
table_data = [["Text", "Actual Label", "TextBlob Polarity", "TextBlob Sentiment", "VADER Compound", "VADER Sentiment"]]

# Lexicon-based approach using TextBlob and VADER
for text, actual_label in data:
    # TextBlob
    blob = TextBlob(text)
    tb_polarity = blob.sentiment.polarity
    
    # Determine label based on polarity score from TextBlob
    if tb_polarity > 0:
        tb_label = 'positive'
    elif tb_polarity < 0:
        tb_label = 'negative'
    else:
        tb_label = 'neutral'

    # VADER
    analyzer = SentimentIntensityAnalyzer()
    vs = analyzer.polarity_scores(text)
    vader_compound = vs['compound']
   
    # Determine label based on compound score from VADER
    if vader_compound > 0.05:
        vader_label = 'positive'
    elif vader_compound < -0.05:
        vader_label = 'negative'
    else:
        vader_label = 'neutral'

    table_data.append([text, actual_label, tb_polarity, tb_label, vader_compound, vader_label])

print(tabulate(table_data, headers="firstrow", tablefmt="plain"))

# Calculate classification report for TextBlob
tb_actual_labels = [label for _, label, _, _, _, _ in table_data[1:]]
tb_predicted_labels = [tb_label for _, _, _, tb_label, _, _ in table_data[1:]]
tb_classification_report = classification_report(tb_actual_labels, tb_predicted_labels, target_names=['negative', 'neutral', 'positive'])

# Calculate classification report for VADER
vader_actual_labels = [label for _, label, _, _, _, _ in table_data[1:]]
vader_predicted_labels = [vader_label for _, _, _, _, _, vader_label in table_data[1:]]
vader_classification_report = classification_report(vader_actual_labels, vader_predicted_labels, target_names=['negative', 'neutral', 'positive'])

# Print classification report for TextBlob
print("\nClassification Report for TextBlob:")
print(tb_classification_report)

# Print classification report for VADER
print("\nClassification Report for VADER:")
print(vader_classification_report)


Text                                                                                                                                                                                                                                                  Actual Label      TextBlob Polarity  TextBlob Sentiment      VADER Compound  VADER Sentiment
good flavor these came securely packed they were fresh and delicious i love these twizzlers                                                                                                                                                           positive                 0.58        positive                        0.9382  positive
this is the same stuff you can buy at the big box stores there is nothing healthy about it it is just carbs and sugars save your money and get something that at least has some taste                                                                 negative                 0.05        positive                        0.2363  positi

# MODEL SELECTION (Machine Learning Approach)

In [30]:
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.svm import SVC
from sklearn.metrics import classification_report

In [31]:
# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(df['Preprocessed_Text'], df['Sentiment'], test_size=0.2, random_state=42)

# Create a TF-IDF vectorizer
tfidf = TfidfVectorizer()

# Fit and transform the training data
X_train_tfidf = tfidf.fit_transform(X_train)

# Transform the testing data
X_test_tfidf = tfidf.transform(X_test)

# Train and evaluate Naive Bayes classifier
nb_classifier = MultinomialNB()
nb_classifier.fit(X_train_tfidf, y_train)
nb_predictions = nb_classifier.predict(X_test_tfidf)
nb_accuracy = accuracy_score(y_test, nb_predictions)
print("Naive Bayes Accuracy:", nb_accuracy)
print(classification_report(y_test, nb_predictions))

# Train and evaluate SVM classifier
svm_classifier = LinearSVC()
svm_classifier.fit(X_train_tfidf, y_train)
svm_predictions = svm_classifier.predict(X_test_tfidf)
svm_accuracy = accuracy_score(y_test, svm_predictions)
print("SVM Accuracy:", svm_accuracy)
print(classification_report(y_test, svm_predictions))

Naive Bayes Accuracy: 0.82
              precision    recall  f1-score   support

    Negative       0.00      0.00      0.00        25
     Neutral       0.00      0.00      0.00        11
    Positive       0.82      1.00      0.90       164

    accuracy                           0.82       200
   macro avg       0.27      0.33      0.30       200
weighted avg       0.67      0.82      0.74       200

SVM Accuracy: 0.845
              precision    recall  f1-score   support

    Negative       0.73      0.32      0.44        25
     Neutral       0.00      0.00      0.00        11
    Positive       0.85      0.98      0.91       164

    accuracy                           0.84       200
   macro avg       0.53      0.43      0.45       200
weighted avg       0.79      0.84      0.80       200



C:\Users\Fahmi\anaconda3\envs\DATAVISUAL\Lib\site-packages\sklearn\metrics\_classification.py:1509: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
C:\Users\Fahmi\anaconda3\envs\DATAVISUAL\Lib\site-packages\sklearn\metrics\_classification.py:1509: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
C:\Users\Fahmi\anaconda3\envs\DATAVISUAL\Lib\site-packages\sklearn\metrics\_classification.py:1509: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(r

# Discussions

Let's delve into the strengths and weaknesses of the chosen sentiment classification models, considering both the experimental outcomes and the attributes of the Amazon Fine Food Reviews dataset.


Lexicon-based Approach Accuracy: 0.804
Naive Bayes Accuracy: 0.82
SVM Accuracy: 0.845

1. Lexicon-based Approach (VADER):
Strengths:

- VADER offers a decent accuracy of 0.804, indicating its ability to capture sentiment to some extent.

- It is relatively simple to implement and can provide quick insights into sentiment without extensive computational resources.

- VADER incorporates sentiment intensity scores, which can offer nuanced sentiment analysis, distinguishing between strong and weak sentiments.

Weaknesses:

- While VADER performs reasonably well, its accuracy (0.804) is lower compared to machine learning models like Naive Bayes and SVM in this context.

- It might struggle with context understanding and nuanced sentiment analysis, especially in cases of sarcasm or complex linguistic expressions.

- VADER's performance heavily relies on the sentiment lexicon it's based on, which might not cover all domain-specific sentiments or newly emerging terms.

2. Naive Bayes Classifier:
- The Naive Bayes classifier achieves an accuracy of 0.82, indicating its effectiveness in capturing sentiment patterns in the dataset.

- Naive Bayes is known for its simplicity, speed, and efficiency, making it suitable for tasks with limited computational resources.

- It can handle a large number of features (words) efficiently, making it scalable to different datasets and domains.

Weaknesses:
- While Naive Bayes performs well, its accuracy (0.82) is slightly lower compared to the SVM classifier in this scenario.

- Naive Bayes assumes independence between features, which may not always hold true in natural language, leading to potential inaccuracies in sentiment prediction.

- It may struggle with handling negations, modifiers, and subtle linguistic nuances.

3. Support Vector Machine (SVM) Classifier:

Strengths:

- The SVM classifier demonstrates the highest accuracy among the models at 0.845, indicating its robustness in capturing sentiment patterns.

- SVMs are effective in handling high-dimensional data and can learn complex decision boundaries, making them suitable for sentiment analysis tasks.

- They perform well in cases where the data is not linearly separable, which can be beneficial for sentiment analysis of nuanced text data.

Weaknesses:
- SVMs can be computationally intensive, especially with large datasets, and may require more resources for training and inference compared to simpler models like Naive Bayes.

- They can be sensitive to the choice of hyperparameters, and tuning these parameters effectively can impact their performance.

- SVMs may not provide as much interpretability compared to simpler models like Naive Bayes or lexicon-based approaches.

In summary, the Naive Bayes classifier and SVM classifier outperform the lexicon-based approach (VADER) in terms of accuracy on the Amazon Fine Food Reviews dataset. While each model has its strengths, such as simplicity for Naive Bayes, robustness for SVM, and quick insights for VADER, they also exhibit weaknesses, such as assumptions of independence for Naive Bayes, computational intensity for SVM, and potential limitations in context understanding for VADER. The choice of model depends on the specific requirements of the sentiment analysis task, including the trade-offs between accuracy, computational resources, interpretability, and the complexity of the dataset.